### PR TITLE
A Litany of TypeChecker Fixes

### PR DESCRIPTION
**Pull request by CodaFi**
### What's in this pull request?

The most important part of this patch is the last few commits.  In theory, we should now be able to handle constructors in most parts of the type checker.

### CODE REVIEW

File: Sources/Mantle/Check.swift [hunk 1]
```diff
@@ -368,14 +368,70 @@ extension TypeChecker where PhaseState == CheckPhaseState {
       self.extendEnvironment([(Name(name: name), patType)])
       return (.variable, type)
     case let .constructor(dataCon, synPats):
+      // Use the data constructor to locate back up the parent so we can
+      // retrieve its argument telescope.
+      guard let data = self.signature.lookupDefinition(dataCon) else {
+        fatalError()
```
Comment (harlanhaskins) [inline Sources/Mantle/Check.swift:374]: These should eventually be replaced with diagnostics, but that shouldn't block this merge.

File: Sources/Mantle/Check.swift [hunk 2]
```diff
@@ -368,14 +368,70 @@ extension TypeChecker where PhaseState == CheckPhaseState {
...
+          fatalError()
+        }
+        let appliedDataConType = self.openContextualType(dataConType, tyConArgs)
+        let (telescope, _) = self.unrollPi(appliedDataConType)
+        let teleSeq = zip((0..<telescope.count).reversed(), telescope)
+        let t = TT.constructor(openDataCon, teleSeq.map { (i, t) in
+          let (nm, _) = t
+          return TT.apply(.variable(Var(nm, UInt(i))), [])
+        })
+        // Now
```
Comment (harlanhaskins) [inline Sources/Mantle/Check.swift]: Finish this?

File: Sources/Mantle/Check.swift (hunk 1 above)
Comment (CodaFi) [reply Sources/Mantle/Check.swift:374]: That one is a proper critical failure of ScopeCheck.  I'll blow through here and have much better primitives for identifying diagnostics versus proper failures.  I also want to try to do the Swift-y thing and crash showing the context of what we were doing and which expression we were type checking.

Comment (harlanhaskins) [reply Sources/Mantle/Check.swift:374]: Ooh, the PrettyStackTrace thing?

**Review by harlanhaskins — APPROVED** [review]

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**